File: CSharpFundamentals/CSharpOOPbasics/PolymorphismEx/VehiclesExtension/Vehicle.cs
using System;

namespace Vehicles
{
    public abstract class Vehicle
    {
        public double FuelConsumption { get; set; }
        public double TankCapacity { get; private set; }
        private double fuelQuantity;

        public double FuelQuantity
        {
            get { return fuelQuantity; }
            set
            {
                if (value > TankCapacity)
                {
                    throw new ArgumentException($"Cannot fit {value} fuel in the tank");
                }
                else if (value <= 0)
                {
                    throw new ArgumentException("Fuel must be a positive number");
                }
                fuelQuantity = value;
            }
        }

        public Vehicle(double fuelQuantity, double fuelConsumption, double tankCapacity)
        {
            TankCapacity = tankCapacity;
            FuelConsumption = fuelConsumption;
            if (tankCapacity < fuelQuantity)
            {
                fuelQuantity = 0;
            }
        }

        public virtual void Drive(double distance)
        {
            if (distance * FuelConsumption <= FuelQuantity)
            {
                Console.WriteLine($"{this.GetType().Name} travelled {distance} km");
                FuelQuantity -= distance * FuelConsumption;
            }
            else
            {
                Console.WriteLine($"{this.GetType().Name} needs refueling");
            }
        }
        public virtual void Refuel(double fuel)
        {
            this.FuelQuantity += fuel;
        }
    }
}

File: JavaFundamentals/JavaOOP/Polymorphism-Ex/src/vehicles_extension/Vehicle.java
package vehicles_extension;

public abstract class Vehicle{
    private Double fuelQuantity;
    private Double fuelConsumption;
    private Double tankCapacity;

    public Vehicle(Double fuelQuantity, Double fuelConsumption, Double tankCapacity) {
        this.setTankCapacity(tankCapacity);
        this.setFuelConsumption(fuelConsumption);
        this.setFuelQuantity(fuelQuantity);
    }

    private void setFuelConsumption(Double fuelConsumption) {
        this.fuelConsumption = fuelConsumption;
    }

    private void setTankCapacity(Double tankCapacity) {
        this.tankCapacity = tankCapacity;
    }

    protected void setFuelQuantity(Double fuelQuantity) {
        if (fuelQuantity < 0 ) {
            throw new IllegalArgumentException("Fuel must be a positive number");
        }
        if (fuelQuantity > this.getTankCapacity()) {
            throw new IllegalArgumentException("Cannot fit fuel in tank");
        }
        this.fuelQuantity = fuelQuantity;
    }

    public Double getFuelQuantity() {
        return this.fuelQuantity;
    }

    public Double getFuelConsumption() {
        return this.fuelConsumption;
    }

    public Double getTankCapacity() {
        return this.tankCapacity;
    }

    public abstract boolean isValidDistance(Double distance);

    public abstract void drive(Double distance);

    public abstract void refuel(Double fuelAmount);
}
